### PR TITLE
cache: enable range, prefix, fromKey key filtering in cache watch

### PR DESCRIPTION
**Pull request by apullo777**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: tests/integration/cache_test.go [hunk 1]
```diff
@@ -45,6 +45,110 @@ func TestCacheWatch(t *testing.T) {
 	testWatch(t, client.KV, c)
 }
 
+func TestCacheWatchSubRange(t *testing.T) {
```
Comment (serathius) [inline tests/integration/cache_test.go:48]: Please merge it into TestWatch

File: tests/integration/cache_test.go [hunk 2]
```diff
@@ -93,44 +197,143 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
...
 		Type: clientv3.EventTypeDelete,
 		Kv: &mvccpb.KeyValue{
 			Key:         []byte("/b"),
 			ModRevision: 5,
 		},
 	}
+	event6PutFoo := &clientv3.Event{
+		Type: clientv3.EventTypePut,
+		Kv: &mvccpb.KeyValue{
+			Key:            []byte("/foo"),
```
Comment (serathius) [inline tests/integration/cache_test.go]: No need to have separate events for testing subranges, you can edit the existing `event1PutA` to make the test work with subranges.

File: tests/integration/cache_test.go [hunk 3]
```diff
@@ -93,44 +197,143 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
...
+			wantEvents: nil,
+		},
+		{
+			name:       "Watch range empty",
+			key:        "",
+			opts:       []clientv3.OpOption{clientv3.WithRange("")},
+			wantEvents: nil,
+		},
+		{
+			name: "Watch range open-end",
```
Comment (serathius) [inline tests/integration/cache_test.go]: Isn't this just `fromKey`? No need to manually put the range.

File: tests/integration/cache_test.go [hunk 4]
```diff
@@ -93,44 +197,143 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
...
+			wantEvents: nil,
+		},
+		{
+			name:       "Watch invalid range [/b, /a)",
+			key:        "/b",
+			opts:       []clientv3.OpOption{clientv3.WithRange("/a")},
+			wantEvents: nil,
+		},
+		{
+			name: "Watch from key /a",
```
Comment (serathius) [inline tests/integration/cache_test.go]: Range from `/a` covers all keys, maybe instead test from `/b`?

File: tests/integration/cache_test.go [hunk 5]
```diff
@@ -93,44 +197,143 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
...
+			key:        "/foo",
+			opts:       []clientv3.OpOption{clientv3.WithFromKey()},
+			wantEvents: []*clientv3.Event{event6PutFoo, event7PutFooBar, event8PutFooBaz},
+		},
+		{
+			name: "Watch from empty key",
+			key:  "",
+			opts: []clientv3.OpOption{clientv3.WithFromKey()},
+			wantEvents: []*clientv3.Event{event1PutA, event2PutB, event3DeleteA, event4PutA, event5DeleteB,
+				event6PutFoo, event7PutFooBar, event8PutFooBaz},
```
Comment (serathius) [inline tests/integration/cache_test.go]: Can you add cases to test `WithPrefix`?

File: tests/integration/cache_test.go [hunk 6]
```diff
@@ -45,6 +45,110 @@ func TestCacheWatch(t *testing.T) {
...
+		want []*clientv3.Event
+	}{
+		{
+			name: "sub-range fully inside cache [/foo/a, /foo/c)",
+			key:  "/foo/a",
+			opts: []clientv3.OpOption{clientv3.WithRange("/foo/c")},
+			want: []*clientv3.Event{event1PutFooA, event2PutFooB, event3DelFooA},
+		},
+		{
+			name: "sub-range partially outside cache [/foo/b, /z)",
```
Comment (serathius) [inline tests/integration/cache_test.go]: We should reject any range that is outside the cached values.

File: tests/integration/cache_test.go [hunk 7]
```diff
@@ -45,6 +45,110 @@ func TestCacheWatch(t *testing.T) {
...
+			key:  "/foo/b",
+			opts: []clientv3.OpOption{clientv3.WithRange("/z")},
+			want: []*clientv3.Event{event2PutFooB, event4PutFooC},
+		},
```
Comment (serathius) [inline tests/integration/cache_test.go]: Please add many more cases, like test WithPrefix and FromKey.

File: tests/integration/cache_test.go (hunk 1 above)
Comment (serathius) [inline tests/integration/cache_test.go]: I recommend to merge this test into the `TestWatch` and handle it:
* For cases where Watch range is within cache range, response from cache and direct watch should match
* For cases where Watch range is outside cache range, cache should return error and direct watch can skip the test.

You can handle skipping by:
* Add `cacheEnabled` argument to `testWatch` function so it caller can make decision based on it
* For cases where Watch range is outside cache range, we can add a field `skipIfCacheEnabled`
* When executing the test cases we can skip cases when `cacheEnabled && tc.skipIfCacheEnabled`

File: tests/integration/cache_test.go (hunk 6 above)
Comment (serathius) [reply tests/integration/cache_test.go]: No, we should expect empty response/cancel.

File: tests/integration/cache_test.go [hunk 8]
```diff
@@ -51,91 +53,228 @@ func TestWatch(t *testing.T) {
...
+			key:        "/foo/b",
+			opts:       []clientv3.OpOption{clientv3.WithRange("/foo/c")},
+			wantEvents: []*clientv3.Event{event2PutFooB, event5DeleteFooB, event7PutFooBar, event8PutFooBaz},
+		},
+		{
+			name:               "Watch range [/foo/c, /z)",
+			key:                "/foo/c",
+			opts:               []clientv3.OpOption{clientv3.WithRange("/z")},
+			wantEvents:         []*clientv3.Event{event6PutFooC, event9PutYoo},
+			skipIfCacheEnabled: true,
```
Comment (serathius) [inline tests/integration/cache_test.go]: We should still test if cache rejects response.

File: tests/integration/cache_test.go [hunk 9]
```diff
@@ -29,20 +29,22 @@ import (
 )
 
 func TestCacheWatch(t *testing.T) {
+	t.Skip("skip until we support cache watch on range")
```
Comment (serathius) [inline tests/integration/cache_test.go]: ???

Comment (serathius) [reply tests/integration/cache_test.go]: We cannot just merge commit that disables tests.

Comment (apullo777) [reply tests/integration/cache_test.go]: Ah, I got it. Just to check I’ve got this right. Since `newWatchEventStream` today returns `ErrUnsupportedWatch` for any sub‐prefix/range, should we update our tests so that any case with Watch on range asserts that the watch is canceled (e.g. resp.Canceled == true) rather than emitting normal events? So perhaps to add a field like `expectCanceled`.

Comment (serathius) [reply tests/integration/cache_test.go]: There is no value of merging code just to test watch. The goal of preparing those tests was to guide your implementation.

File: tests/integration/cache_test.go (hunk 8 above)
Comment (apullo777) [reply tests/integration/cache_test.go]: Ok. I will modify `skipIfCacheEnabled` to be `expectCanceled` and change the validate logic like below:

```
	t.Log("Validate")
	for i, tc := range tcs {
		i, tc := i, tc
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			resp, ok := <-watches[i]
			if !ok {
				t.Fatalf("unexpected closed watch")
			}
			if resp.Canceled {
				if tc.expectCanceled {
					// invalid range or cache-only out of range key
					return
				}
				t.Fatalf("unexpected canceled watch: %+v", resp)
			}
			// We ignore the first WatchResponse which is the CreatedNotify (len(resp.Events) == 0) and read the real payload that follows.
			events, _ := readEvents(watches[i])
			if diff := cmp.Diff(tc.wantEvents, events); diff != "" {
				t.Errorf("unexpected events (-want +got):\n%s", diff)
			}
		})
	}
```
and then also add `clientv3.WithCreatedNotify()` so that the initial `<-watches[i] ` unblock deterministically:

```
		watches[i] = watcher.Watch(
			ctx,
			tc.key,
			append(tc.opts, clientv3.WithCreatedNotify())...,
		)
```

File: tests/integration/cache_test.go (hunk 9 above)
Comment (apullo777) [reply tests/integration/cache_test.go]: Sorry. Totally misunderstood the purpose of this PR. I will continue work on this PR to do the implementation using these tests.

**Comment by serathius** [comment]
Please fix conflict.

File: cache/cache.go [hunk 10]
```diff
@@ -103,6 +103,23 @@ func (c *Cache) Watch(ctx context.Context, key string, opts ...clientv3.OpOption
...
 	requestedRev := op.Rev()
 
+	startKey := []byte(key)
+	endKey := op.RangeBytes() // nil → single key, {0} → fromKey
+
+	prefix := []byte(c.prefix)
+	startIn := bytes.HasPrefix(startKey, prefix)
+	endIn := len(endKey) == 0 || bytes.HasPrefix(endKey, prefix)
+	openEnded := len(endKey) == 1 && endKey[0] == 0
+	invalid := len(endKey) > 0 && !openEnded && bytes.Compare(startKey, endKey) >= 0
```
Comment (serathius) [inline cache/cache.go]: I said to not use `prefix(a)` semantic, but use `range(a,b)`

File: cache/cache.go [hunk 11]
```diff
@@ -105,8 +105,35 @@ func (c *Cache) Watch(ctx context.Context, key string, opts ...clientv3.OpOption
 	op := clientv3.OpGet(key, opts...)
 	startRev := op.Rev()
 
-	// TODO: Support watch on subprefix and single key & arbitrary startRev support once we guarantee gap-free replay.
-	if key != c.prefix || !clientv3.IsOptsWithPrefix(opts) || startRev != 0 {
+	// TODO: Support arbitrary startRev support once we guarantee gap-free replay.
+	if  startRev != 0 {
```
Comment (serathius) [inline cache/cache.go]: This function is too long, move watch validation to separate function.

File: cache/cache.go [hunk 12]
```diff
@@ -226,3 +221,29 @@ func readWatchChannel(
 	}
 	return nil
 }
+
+func (c *Cache) validateWatch(key string, opts ...clientv3.OpOption) (cancelChan clientv3.WatchChan, pred KeyPredicate) {
```
Comment (serathius) [inline cache/cache.go]: `validateWatch` should validate watch and return error if failed, not return `(cancelChan clientv3.WatchChan, pred KeyPredicate)`

File: cache/predicate.go [hunk 13]
```diff
@@ -23,3 +25,16 @@ func (prefix Prefix) Match(key []byte) bool {
 	prefixLen := len(prefix)
 	return len(key) >= prefixLen && string(key[:prefixLen]) == string(prefix)
 }
+
+func rangePred(start, end []byte) func(k []byte) bool {
+	return func(k []byte) bool {
+		if bytes.Compare(k, start) < 0 {
+			return false
+		}
+		// no end, or the sentinel {0}  => accept everything ≥ start
```
Comment (serathius) [inline cache/predicate.go]: Don't use sentinel, create separate predicate.

File: cache/cache.go [hunk 14]
```diff
@@ -226,3 +221,29 @@ func readWatchChannel(
...
+
+	switch {
+	case len(endKey) == 0:
+		pred = func(k []byte) bool { return bytes.Equal(k, startKey) }
+	case openEnd:
+		pred = func(k []byte) bool { return bytes.Compare(k, startKey) >= 0 }
+	default:
+		pred = rangePred(startKey, endKey)
+	}
+	return nil, pred
```
Comment (serathius) [inline cache/cache.go]: Move code to create predicate to dedicated function in `cache/predicate.go`

File: cache/cache.go [hunk 15]
```diff
@@ -102,26 +102,21 @@ func (c *Cache) Watch(ctx context.Context, key string, opts ...clientv3.OpOption
...
-		return responseChan
+	cancelChan, pred := c.validateWatch(key, opts...)
+	if cancelChan != nil {
+		return cancelChan
 	}
 
-	w := newWatcher(c.cfg.PerWatcherBufferSize, func(k []byte) bool { return strings.HasPrefix(string(k), key) })
-	c.demux.Register(w, startRev)
+	w := newWatcher(c.cfg.PerWatcherBufferSize, pred)
+	c.demux.Register(w, 0)
```
Comment (serathius) [inline cache/cache.go]:
```suggestion
	startRev := 0
	c.demux.Register(w, startRev)
```

Comment (serathius) [reply cache/cache.go]: Setting here a zero obfuscates the argument.

File: cache/cache.go (hunk 12 above)
Comment (MadhavJivrajani) [reply cache/cache.go]: +1
Interpret the error at the call sight and then close the channel like you do currently

File: tests/integration/cache_test.go (hunk 8 above)
Comment (apullo777) [reply tests/integration/cache_test.go]: I found that TestCacheLaggingWatcher/all_event_fit was failing with:

```
=== RUN   TestCacheLaggingWatcher/all_event_fit
    cache_test.go:367: gotEvents=0, wantEvents=9
    cache_test.go:373: closed=true, wantClosed=false
```
because the unbuffered channel blocked on the {Created:true} send, and never read real events.

I noticed that client/v3/watch.go do  `make(chan WatchResponse, 1)` before sending the Created notification. So I tried to follow this pattern to switched the watch’s response channel to a one-slot buffer:  
`responseChan := make(chan clientv3.WatchResponse, 1)`

We could also drop the “created” notification entirely, but I need to figure out how to do that.

File: cache/cache.go [hunk 16]
```diff
@@ -105,23 +104,24 @@ func (c *Cache) Watch(ctx context.Context, key string, opts ...clientv3.OpOption
...
 	c.demux.Register(w, startRev)
 
-	responseChan := make(chan clientv3.WatchResponse)
+	responseChan := make(chan clientv3.WatchResponse, 1)
 	c.waitGroup.Add(1)
 	go func() {
 		defer c.waitGroup.Done()
 		defer close(responseChan)
 		defer c.demux.Unregister(w)
+		responseChan <- clientv3.WatchResponse{Created: true}
```
Comment (serathius) [inline cache/cache.go]: Create response should only be send if client requests it.

File: cache/cache.go [hunk 17]
```diff
@@ -175,7 +175,7 @@ func serveWatchEvents(ctx context.Context, watchCtx *watchCtx) {
 	backoff := watchCtx.backoffStart
 	for {
 		opts := []clientv3.OpOption{
-			clientv3.WithPrefix(),
+			clientv3.WithFromKey(),
```
Comment (serathius) [inline cache/cache.go]: This should not be changed.

Comment (apullo777) [reply cache/cache.go]: Ok. I will add a helper so that testWatch can use it to filter wantEvents by cache's prefix.

Comment (serathius) [reply cache/cache.go]: I don't understand.

Comment (apullo777) [reply cache/cache.go]: Sorry, I did not explain it well. We have a generic table-driven helper `testWatch` that runs the same scenarios against both a raw client watch and our cache-backed watch, and in the table we includes an event on "/yoo" (event 9) that is outside cache's prefix "/foo" (we called `cache.New(cli, "/foo", …)` in `TestCacheWatch`).  I just found that I did not design a mechanism in `testWatch` to filter for test-cases that originally expected to see the “/yoo” event for `TestWatch` so the cache tests will constantly fail with “missing event” with `WithPrefix()`.

Comment (serathius) [reply cache/cache.go]: Ok, please be clear next time to avoid confusion. No need to be so detailed. 

"Ok, will also need to update tests scenarios because they depend on this" is enough.

File: cache/cache.go [hunk 18]
```diff
@@ -226,3 +230,37 @@ func readWatchChannel(
...
+
+	if isRange && bytes.Compare(endKey, startKey) <= 0 {
+		return nil, ErrUnsupportedWatch
+	}
+
+	if c.prefix != "" && isRange {
+		prefixStart := []byte(c.prefix)
+		prefixEndStr := clientv3.GetPrefixRangeEnd(c.prefix)
+		prefixEnd := []byte(prefixEndStr)
+		endBound := endKey
```
Comment (serathius) [inline cache/cache.go]:
```suggestion
```

File: cache/cache.go [hunk 19]
```diff
@@ -226,3 +230,37 @@ func readWatchChannel(
...
+
+		if bytes.Compare(startKey, prefixStart) < 0 ||
+			bytes.Compare(startKey, prefixEnd) >= 0 ||
+			bytes.Compare(endBound, prefixStart) <= 0 ||
+			bytes.Compare(endBound, prefixEnd) > 0 {
+			return nil, ErrUnsupportedWatch
+		}
+	}
```
Comment (serathius) [inline cache/cache.go]: Validating ranges should be a separate function and have dedicated unit tests.

File: tests/integration/cache_test.go [hunk 20]
```diff
@@ -51,93 +52,258 @@ func TestWatch(t *testing.T) {
...
+			key:        "/foo/b",
+			opts:       []clientv3.OpOption{clientv3.WithRange("/foo/c")},
+			wantEvents: []*clientv3.Event{event2PutFooB, event5DeleteFooB, event7PutFooBar, event8PutFooBaz},
+		},
+		{
+			name:               "Watch valid range [/foo/c, /z)",
+			key:                "/foo/c",
+			opts:               []clientv3.OpOption{clientv3.WithRange("/z")},
+			wantEvents:         []*clientv3.Event{event6PutFooC, event9PutYoo},
+			skipIfCacheEnabled: true,
```
Comment (serathius) [inline tests/integration/cache_test.go]: Instead of testing direct watch and skipping testing cache, you should do reverse. Skip testing watch, and test that cache returns error for cache outside of cache prefix.

File: tests/integration/cache_test.go [hunk 21]
```diff
@@ -51,93 +52,258 @@ func TestWatch(t *testing.T) {
...
+				return
+			}
 			events, _ := readEvents(watches[i])
-			if diff := cmp.Diff(tc.wantEvents, events); diff != "" {
+			filteredWant := tc.wantEvents
+			if cp, ok := watcher.(interface{ Prefix() string }); ok {
+				if p := cp.Prefix(); p != "" {
+					filteredWant = filterByPrefix(filteredWant, p)
+				}
+			}
```
Comment (serathius) [inline tests/integration/cache_test.go]: There should be no events filtered. 

Don't know what it does, but it's definitely wrong. Please remove.

File: tests/integration/cache_test.go [hunk 22]
```diff
@@ -144,6 +252,73 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
 	}
 }
 
+func TestCacheRejectsInvalidWatchOptions(t *testing.T) {
```
Comment (serathius) [inline tests/integration/cache_test.go]: This test should be merged with `TestCacheRejectsUnsupportedWatch`. We should have single test to track what options are supported/work.

File: cache/cache.go [hunk 23]
```diff
@@ -226,3 +230,59 @@ func readWatchChannel(
...
+			bytes.Compare(endKey, prefixStart) <= 0 ||
+			bytes.Compare(endKey, prefixEnd) > 0 {
+			return ErrUnsupportedWatch
+		}
+	}
+
+	// single-key watch must lie within [prefixStart, prefixEnd)
+	if bytes.Compare(startKey, prefixStart) < 0 || bytes.Compare(startKey, prefixEnd) >= 0 {
+		return ErrUnsupportedWatch
+	}
```
Comment (serathius) [inline cache/cache.go]: This should be switch because cases are mutually exclusive

File: cache/cache.go [hunk 24]
```diff
@@ -165,6 +165,10 @@ func (c *Cache) WaitReady(ctx context.Context) error {
 	}
 }
 
+func (c *Cache) Prefix() string {
```
Comment (serathius) [inline cache/cache.go]: This is no longer used.

File: cache/cache.go [hunk 25]
```diff
@@ -226,3 +230,59 @@ func readWatchChannel(
...
+
+	if err := c.validateWatchRange(startKey, endKey); err != nil {
+		return nil, err
+	}
+	return KeyPredForRange(startKey, endKey), nil
+}
+
+func (c *Cache) validateWatchRange(startKey, endKey []byte) error {
+	isFrom := len(endKey) == 1 && endKey[0] == 0
+	isRange := endKey != nil && !isFrom
```
Comment (serathius) [inline cache/cache.go]: What about case where `endKey` is empty?

File: cache/cache.go [hunk 26]
```diff
@@ -226,3 +230,59 @@ func readWatchChannel(
...
+	prefixStart := []byte(c.prefix)
+	prefixEnd := []byte(clientv3.GetPrefixRangeEnd(c.prefix))
+
+	if isRange && bytes.Compare(endKey, startKey) <= 0 {
```
Comment (serathius) [inline cache/cache.go]: This code for testing range should be next to line 273 where there is other statement to test range.

File: cache/cache.go [hunk 27]
```diff
@@ -226,3 +230,59 @@ func readWatchChannel(
...
+		return ErrUnsupportedWatch
+	}
+
+	if c.prefix == "" {
```
Comment (serathius) [inline cache/cache.go]: Don't like that you are treating `c.prefix == ""` as special case. Will rest of the validation fail?

File: cache/cache_test.go [hunk 28]
```diff
@@ -0,0 +1,199 @@
...
+			end:     []byte{0},
+			wantErr: false,
+		},
+		{
+			name:    "prefix m: fromKey above prefix end error",
+			prefix:  "m",
+			start:   []byte("z"),
+			end:     []byte{0},
+			wantErr: true,
+		},
```
Comment (serathius) [inline cache/cache_test.go]: hardcoding cases based on byte representation of client options like `fromKey` makes the test less readable. The test case should be:

```
		{
			name:    "prefix m: fromKey above prefix end error",
			cachePrefix:  "m",
			watchKey: "z",
			option: clientv3.WithFromKey(),
			wantErr: true,
		},
```

And the test then would be:
```
			op := clientv3.OpGet(c.key, c.opt)
			err := dummyCache.validateWatchRange([]byte(c.key), op.RangeBytes())
			if gotErr := err != nil; gotErr != c.wantErr {
				t.Fatalf("validateWatchRange(%q, %v) = %v, wantErr=%v",
					c.prefix, c.start, err, c.wantErr)
			}
```

File: cache/cache_test.go [hunk 29]
```diff
@@ -0,0 +1,199 @@
...
+		{
+			name:    "root: reverse range error",
+			prefix:  "",
+			start:   []byte("b"),
+			end:     []byte("a"),
+			wantErr: true,
+		},
+		{
+			name:    "prefix m: range below start error",
```
Comment (serathius) [inline cache/cache_test.go]: `prefix m:` doesn't inform anything about the test. Also client we would not normally use `m`, tests should cover plausible arguments.


See example:
```
		{
			name:    "range encompassing cache prefix returns error",
			key:   []byte("/a"),
			opt:   clientv3.WithRange("/z")
			cachePrefix:  "/foo",
			wantErr: true,
		},
```

File: cache/cache.go (hunk 27 above)
Comment (apullo777) [reply cache/cache.go]: I found it difficult to handle the empty string prefix case, which will get us a non-nil, zero-length []byte{""} (and then `GetPrefixRangeEnd("")` give us a `prefixEnd` with []byte{0}). An option I have in mind is to translate its `prefixStart` & `prefixEnd` to represent open-ended. Or we can hide this special case to be a check before we do prefix validation like below:

```
	prefixStart := []byte(c.prefix)
	prefixEnd := []byte(clientv3.GetPrefixRangeEnd(c.prefix))

	switch {
	case len(endKey) == 0: // single key
		if c.prefix != "" &&
			(bytes.Compare(startKey, prefixStart) < 0 ||
				bytes.Compare(startKey, prefixEnd) >= 0) {
			return ErrUnsupportedWatch
		}

	case len(endKey) == 1 && endKey[0] == 0: //fromKey
		if c.prefix != "" &&
			bytes.Compare(startKey, prefixEnd) >= 0 {
			return ErrUnsupportedWatch
		}

	default: // range
		if bytes.Compare(endKey, startKey) <= 0 {
			return ErrUnsupportedWatch
		}
		if c.prefix != "" &&
			(bytes.Compare(startKey, prefixStart) < 0 ||
				bytes.Compare(startKey, prefixEnd) >= 0 ||
				bytes.Compare(endKey, prefixStart) <= 0 ||
				bytes.Compare(endKey, prefixEnd) > 0) {
			return ErrUnsupportedWatch
		}
	}
```

Comment (serathius) [reply cache/cache.go]: That's better

File: cache/cache_test.go [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+		},
+		{
+			name:        "range crossing prefixEnd returns error",
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/z")},
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "range crossing prefixEnd returns error",
```
Comment (serathius) [inline cache/cache_test.go]: This test is duplicated

File: cache/cache_test.go [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+		},
+		{
+			name:        "empty prefix",
+			watchKey:    "",
+			opts:        []clientv3.OpOption{clientv3.WithPrefix()},
+			cachePrefix: "",
+			wantErr:     false,
+		},
+		{
+			name:        "empty prefix with cachePrefix",
```
Comment (serathius) [inline cache/cache_test.go]: This should return error, cache will return only objects with "/foo" prefix, while the watcher wants all objects.

File: cache/cache_test.go [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+		},
+		{
+			name:        "fromKey open‑ended",
+			watchKey:    "/a",
+			opts:        []clientv3.OpOption{clientv3.WithFromKey()},
+			cachePrefix: "",
+			wantErr:     false,
+		},
+		{
+			name:        "fromKey starting at prefix start",
```
Comment (serathius) [inline cache/cache_test.go:157]: This should return error. Watch from "/foo" allows for keys like "/z" that are not within prefix.

File: cache/cache_test.go [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithFromKey()},
+			cachePrefix: "/foo",
+			wantErr:     false,
+		},
+		{
+			name:        "fromKey starting below prefixEnd",
```
Comment (serathius) [inline cache/cache_test.go:164]: This should return error.

File: cache/cache_test.go [hunk 34]
```diff
@@ -0,0 +1,203 @@
...
+
+	for _, c := range tests {
+		t.Run(c.name, func(t *testing.T) {
+			dummyCache := &Cache{prefix: c.cachePrefix}
+			var op clientv3.Op
+			if len(c.opts) == 0 {
+				op = clientv3.OpGet(c.watchKey)
+			} else {
+				op = clientv3.OpGet(c.watchKey, c.opts...)
+			}
```
Comment (serathius) [inline cache/cache_test.go]:
```suggestion
			op := clientv3.OpGet(c.watchKey, c.opts...)
```

File: cache/cache_test.go (hunk 33 above)
Comment (apullo777) [reply cache/cache_test.go:164]: Thanks! Just want to check if I got this right:

- A root cache (prefix=="") will accept any valid etcd watch.
- A shard cache (e.g. prefix="/foo") can only accept watches whose entire [startKey,endKey) lies within "/foo"…"/fop",. And with `WithFromKey()` we should reject a watch starting anywhere below "/foo" even though that will also return events within "/foo"…"/fop".

Comment (serathius) [reply cache/cache_test.go:164]: Yes, there should not be any events missing in watch, as so watch can only be opened on cache that encompasses its range. For WithFromKey() it means it only works for cache with prefix "".

Comment (apullo777) [reply cache/cache_test.go:164]: Ok. Then I should also modify integration tests to make sure cache rejecting WithFromKey & empty WithPrefix.

Comment (serathius) [reply cache/cache_test.go:164]: Sounds good

File: cache/cache.go [hunk 35]
```diff
@@ -226,3 +226,54 @@ func readWatchChannel(
...
+	endKey := op.RangeBytes() // nil = single key, {0}=FromKey, else explicit range
+
+	if err := c.validateWatchRange(startKey, endKey); err != nil {
+		return nil, err
+	}
+	return KeyPredForRange(startKey, endKey), nil
+}
+
+func (c *Cache) validateWatchRange(startKey, endKey []byte) error {
+	shard := c.prefix != ""
```
Comment (serathius) [inline cache/cache.go]: What is shard? There is no such concept in etcd. Not it was ever mentioned in our discussions. Please revert to previous code.

File: cache/cache.go [hunk 36]
```diff
@@ -226,3 +226,54 @@ func readWatchChannel(
...
+	isSingleKey := len(endKey) == 0
+	isFromKey := len(endKey) == 1 && endKey[0] == 0
+
+	switch {
+	case isSingleKey:
+		if shard &&
+			(bytes.Compare(startKey, prefixStart) < 0 ||
+				bytes.Compare(startKey, prefixEnd) >= 0) {
+			return ErrUnsupportedWatch
+		}
```
Comment (serathius) [inline cache/cache.go]:
```suggestion
		if c.prefix == "" {
			return nil
		}
		if bytes.Compare(startKey, prefixStart) < 0 || bytes.Compare(startKey, prefixEnd) >= 0 {
			return ErrUnsupportedWatch
		}
		return nil
```

File: cache/cache.go [hunk 37]
```diff
@@ -226,3 +226,54 @@ func readWatchChannel(
...
+
+	case isFromKey:
+		if shard {
+			return ErrUnsupportedWatch
+		}
```
Comment (serathius) [inline cache/cache.go]:
```suggestion
		if c.prefix != "" {
			return ErrUnsupportedWatch
		}
		return nil
```

File: cache/cache.go [hunk 38]
```diff
@@ -226,3 +226,54 @@ func readWatchChannel(
...
+	default:
+		if bytes.Compare(endKey, startKey) <= 0 {
+			return ErrUnsupportedWatch
+		}
+		if shard &&
+			(bytes.Compare(startKey, prefixStart) < 0 ||
+				bytes.Compare(endKey, prefixEnd) > 0) {
+			return ErrUnsupportedWatch
+		}
+	}
```
Comment (serathius) [inline cache/cache.go]:
```suggestion
		if bytes.Compare(endKey, startKey) <= 0 {
			return ErrUnsupportedWatch
		}
		if c.prefix == "" {
			return nil
		}
		if bytes.Compare(startKey, prefixStart) < 0 || bytes.Compare(endKey, prefixEnd) > 0 {
			return ErrUnsupportedWatch
		}
		return nil
	}
```

File: cache/cache_test.go [hunk 39]
```diff
@@ -0,0 +1,191 @@
...
+	}
+
+	tests := []tc{
+		{
+			name:        "single key",
+			watchKey:    "/a",
+			cachePrefix: "",
+			wantErr:     false,
+		},
+		{
```
Comment (serathius) [inline cache/cache_test.go:39]:
```suggestion
		{
			name:        "prefix single key",
			watchKey:    "/foo/a",
			cachePrefix: "/foo",
			wantErr:     false,
		},
		{
```

File: cache/cache_test.go [hunk 40]
```diff
@@ -0,0 +1,191 @@
...
+			name:        "empty range returns error",
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/foo")},
+			cachePrefix: "",
+			wantErr:     true,
+		},
+		{
+			name:        "range strating below cache prefix returns error",
```
Comment (serathius) [inline cache/cache_test.go]:
```suggestion
			name:        "range starting below cache prefix returns error",
```

File: cache/cache_test.go [hunk 41]
```diff
@@ -0,0 +1,191 @@
...
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "range ending above cache prefix returns error",
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/z")},
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
```
Comment (serathius) [inline cache/cache_test.go]:
```suggestion
```

This test duplicates `range crossing prefixEnd returns error`

File: cache/cache_test.go [hunk 42]
```diff
@@ -0,0 +1,191 @@
...
+		{
+			name:        "empty prefix",
+			watchKey:    "",
+			opts:        []clientv3.OpOption{clientv3.WithPrefix()},
+			cachePrefix: "",
+			wantErr:     false,
+		},
+		{
+			name:        "empty prefix with cachePrefix",
```
Comment (serathius) [inline cache/cache_test.go]:
```suggestion
			name:        "empty prefix with cachePrefix returns error",
```

File: cache/cache_test.go [hunk 43]
```diff
@@ -0,0 +1,191 @@
...
+			watchKey:    "",
+			opts:        []clientv3.OpOption{clientv3.WithPrefix()},
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "prefix watch matches cachePrefi xexactly",
```
Comment (serathius) [inline cache/cache_test.go]:
```suggestion
			name:        "prefix watch matches cachePrefix exactly",
```

File: tests/integration/cache_test.go [hunk 44]
```diff
@@ -34,7 +34,7 @@ func TestCacheWatch(t *testing.T) {
 	t.Cleanup(func() { clus.Terminate(t) })
 	client := clus.Client(0)
 
-	c, err := cache.New(client, "/", cache.WithHistoryWindowSize(32))
+	c, err := cache.New(client, "", cache.WithHistoryWindowSize(32))
```
Comment (serathius) [inline tests/integration/cache_test.go]: We wanted to test cache opened on prefix "/foo". Why change that?

File: tests/integration/cache_test.go [hunk 45]
```diff
@@ -56,84 +56,192 @@ func TestWatch(t *testing.T) {
...
+		Type: clientv3.EventTypePut,
+		Kv: &mvccpb.KeyValue{
+			Key:            []byte("/foo/baz"),
+			Value:          []byte("z"),
+			CreateRevision: 8,
+			ModRevision:    8,
+			Version:        1,
+		},
+	}
+	event9PutFooYoo := &clientv3.Event{
```
Comment (serathius) [inline tests/integration/cache_test.go:133]: Can you add one write with key without `/foo` prefix?

File: tests/integration/cache_test.go (hunk 44 above)
Comment (apullo777) [reply tests/integration/cache_test.go]: I removed it because now our cache will reject events outside cache prefix, and testEatch will compare their results end to end. Perhaps I should find a way to handle these different expectations.

Comment (serathius) [reply tests/integration/cache_test.go]: I think we need separate tests when cache prefix is "" or "/foo". However this is not trivial to implement. Let's skip it for now and just test "/foo".

**Review by serathius — APPROVED** [review]
